// @nozbe/watermelondb/decorators/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "action", {
  enumerable: true,
  get: function get() {
    return _action.default;
  }
});
Object.defineProperty(exports, "children", {
  enumerable: true,
  get: function get() {
    return _children.default;
  }
});
Object.defineProperty(exports, "json", {
  enumerable: true,
  get: function get() {
    return _json.default;
  }
});
Object.defineProperty(exports, "nochange", {
  enumerable: true,
  get: function get() {
    return _nochange.default;
  }
});
Object.defineProperty(exports, "field", {
  enumerable: true,
  get: function get() {
    return _field.default;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _date.default;
  }
});
Object.defineProperty(exports, "text", {
  enumerable: true,
  get: function get() {
    return _text.default;
  }
});
Object.defineProperty(exports, "readonly", {
  enumerable: true,
  get: function get() {
    return _readonly.default;
  }
});
Object.defineProperty(exports, "lazy", {
  enumerable: true,
  get: function get() {
    return _lazy.default;
  }
});
Object.defineProperty(exports, "relation", {
  enumerable: true,
  get: function get() {
    return _relation.default;
  }
});
Object.defineProperty(exports, "immutableRelation", {
  enumerable: true,
  get: function get() {
    return _immutableRelation.default;
  }
});

var _action = _interopRequireDefault(require("./action"));

var _children = _interopRequireDefault(require("./children"));

var _json = _interopRequireDefault(require("./json"));

var _nochange = _interopRequireDefault(require("./nochange"));

var _field = _interopRequireDefault(require("./field"));

var _date = _interopRequireDefault(require("./date"));

var _text = _interopRequireDefault(require("./text"));

var _readonly = _interopRequireDefault(require("./readonly"));

var _lazy = _interopRequireDefault(require("./lazy"));

var _relation = _interopRequireDefault(require("./relation"));

var _immutableRelation = _interopRequireDefault(require("./immutableRelation"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
