// @nozbe/watermelondb/utils/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getPreciseTime", {
  enumerable: true,
  get: function get() {
    return _devMeasureTime.getPreciseTime;
  }
});
Object.defineProperty(exports, "devMeasureTime", {
  enumerable: true,
  get: function get() {
    return _devMeasureTime.devMeasureTime;
  }
});
Object.defineProperty(exports, "devMeasureTimeAsync", {
  enumerable: true,
  get: function get() {
    return _devMeasureTime.devMeasureTimeAsync;
  }
});
Object.defineProperty(exports, "randomId", {
  enumerable: true,
  get: function get() {
    return _randomId.default;
  }
});
Object.defineProperty(exports, "makeDecorator", {
  enumerable: true,
  get: function get() {
    return _makeDecorator.default;
  }
});
Object.defineProperty(exports, "ensureSync", {
  enumerable: true,
  get: function get() {
    return _ensureSync.default;
  }
});
Object.defineProperty(exports, "invariant", {
  enumerable: true,
  get: function get() {
    return _invariant.default;
  }
});
Object.defineProperty(exports, "logError", {
  enumerable: true,
  get: function get() {
    return _logError.default;
  }
});
Object.defineProperty(exports, "logger", {
  enumerable: true,
  get: function get() {
    return _logger.default;
  }
});
Object.defineProperty(exports, "connectionTag", {
  enumerable: true,
  get: function get() {
    return _connectionTag.default;
  }
});

var _devMeasureTime = require("./devMeasureTime");

var _randomId = _interopRequireDefault(require("./randomId"));

var _makeDecorator = _interopRequireDefault(require("./makeDecorator"));

var _ensureSync = _interopRequireDefault(require("./ensureSync"));

var _invariant = _interopRequireDefault(require("./invariant"));

var _logError = _interopRequireDefault(require("./logError"));

var _logger = _interopRequireDefault(require("./logger"));

var _connectionTag = _interopRequireDefault(require("./connectionTag"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
